internal/xds/util.go

```diff
@@ -38,3 +60,157 @@ func ClusterLoadAssignmentName(service types.NamespacedName, portName string) st
+	case *envoy_config_listener_v3.Listener:
+		for i := range msg.ListenerFilters {
+			msg.ListenerFilters[i] = Fixup(msg.ListenerFilters[i]).(*envoy_config_listener_v3.ListenerFilter)
+		}
+
+		for _, chain := range msg.FilterChains {
+			for i := range chain.Filters {
+				chain.Filters[i].ConfigType = &envoy_config_listener_v3.Filter_TypedConfig{
+					TypedConfig: Fixup(chain.Filters[i].GetTypedConfig()).(*any.Any),
+				}
+			}
+
+			if t := chain.GetTransportSocket(); t != nil {
+				chain.TransportSocket.ConfigType = &envoy_config_core_v3.TransportSocket_TypedConfig{
+					TypedConfig: Fixup(t.GetTypedConfig()).(*any.Any),
+				}
+			}
+		}
+
+		return in
+
+	case *any.Any:
+		anyval := ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(msg, &anyval); err != nil {
+			panic(fmt.Sprintf("failed to unmarshal %T: %s", msg, err.Error()))
+		}
+
+		return protobuf.MustMarshalAny(Fixup(anyval.Message))
+
+	case *envoy_config_listener_v3.Filter:
+		if c := msg.GetTypedConfig(); c != nil {
+			msg.ConfigType = &envoy_config_listener_v3.Filter_TypedConfig{
+				TypedConfig: Fixup(c).(*any.Any),
+			}
+		}
+
+		return msg
+
+	case *envoy_config_listener_v3.ListenerFilter:
+		if c := msg.GetTypedConfig(); c != nil {
+			msg.ConfigType = &envoy_config_listener_v3.ListenerFilter_TypedConfig{
+				TypedConfig: Fixup(c).(*any.Any),
+			}
+		}
+
+		return msg
+
+	case *envoy_config_cluster_v3.Cluster:
+		if e := msg.GetEdsClusterConfig(); e != nil {
+			e.GetEdsConfig().ResourceApiVersion = envoy_config_core_v3.ApiVersion_V3
+			e.GetEdsConfig().GetApiConfigSource().TransportApiVersion = envoy_config_core_v3.ApiVersion_V3
+		}
+
+		if t := msg.GetTransportSocket(); t != nil {
+			t.ConfigType = &envoy_config_core_v3.TransportSocket_TypedConfig{
+				TypedConfig: Fixup(t.GetTypedConfig()).(*any.Any),
+			}
+		}
+
+		return msg
+
+	case *envoy_config_route_v3.RouteConfiguration:
+		for _, v := range msg.GetVirtualHosts() {
+			for _, r := range v.GetRoutes() {
+				for name, conf := range r.GetTypedPerFilterConfig() {
+					r.TypedPerFilterConfig[name] = Fixup(conf).(*any.Any)
+				}
+			}
+		}
+
+		return msg
+
+	case *envoy_extensions_transport_sockets_tls_v3.Secret:
+		return msg
+
+	case *envoy_config_endpoint_v3.ClusterLoadAssignment:
+		return msg
+
+	case *envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager:
+		v3 := envoy_extensions_filters_network_http_connection_manager_v3.HttpConnectionManager{}
+		protobuf.MustConvertTo(msg, &v3)
+
+		if r := v3.GetRds(); r != nil {
+			r.GetConfigSource().ResourceApiVersion = envoy_config_core_v3.ApiVersion_V3
+			r.GetConfigSource().GetApiConfigSource().TransportApiVersion = envoy_config_core_v3.ApiVersion_V3
+		}
+
+		for _, f := range v3.HttpFilters {
+			if c := f.GetTypedConfig(); c != nil {
+				f.ConfigType = &envoy_extensions_filters_network_http_connection_manager_v3.HttpFilter_TypedConfig{
+					TypedConfig: Fixup(c).(*any.Any),
+				}
+			}
+		}
+
```

Comment:
I think you need something like this in here for the access log config, it was still coming through as v2 without:

```
for i, a := range v3.AccessLog {
        v3.AccessLog[i].ConfigType = &envoy_config_accesslog_v3.AccessLog_TypedConfig{
	        TypedConfig: Fixup(a.GetTypedConfig()).(*any.Any),
	}
}
```

Reply:
fixed in HCM typeswitch